### PR TITLE
[FLINK-33410] Implement type inference for OVER function

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

It implements type inference for `OVER`, `STREAM_RECORD` and `THROW_EXCEPTION` functions.

## Verifying this change
* all available tests pass
* added tests for `OVER` type inference

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 1]
```diff
@@ -2057,35 +2057,8 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
...
-                    .build();
-
-    public static final BuiltInFunctionDefinition CURRENT_RANGE =
-            BuiltInFunctionDefinition.newBuilder()
-                    .name("currentRange")
-                    .kind(OTHER)
-                    .outputTypeStrategy(TypeStrategies.MISSING)
-                    .build();
-
-    public static final BuiltInFunctionDefinition CURRENT_ROW =
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2084]: With a symbol. Those were actually not really functions, but a poor imitation of a symbol.

If I implement them properly as symbols I can also do the validation and check for those in `OverTypeStrategy`.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/expressions/PlannerExpressionConverter.scala [hunk 2]
```diff
@@ -79,12 +79,6 @@ class PlannerExpressionConverter private extends ApiExpressionVisitor[PlannerExp
         val windowReference = translateWindowReference(children.head)
         return WindowEnd(windowReference)
 
-      case THROW_EXCEPTION =>
```
Comment (jnh5y) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/expressions/PlannerExpressionConverter.scala:82]: Interesting.  Seems reasonable to remove this.  

Also, seems unrelated.

Do we need the THROW_EXCEPTION function?  Or can it be removed in general?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2084]: Sounds like separate follow-on work.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/expressions/PlannerExpressionConverter.scala (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/expressions/PlannerExpressionConverter.scala:82]: We still have it. I guess you are mislead that it is not listed in `BuiltinFunctionDefinitions` but `InternalFunctionDefinitions`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 3]
```diff
@@ -2302,7 +2302,8 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
             BuiltInFunctionDefinition.newBuilder()
                     .name("streamRecordTimestamp")
                     .kind(OTHER)
-                    .outputTypeStrategy(TypeStrategies.MISSING)
+                    .inputTypeStrategy(NO_ARGS)
+                    .outputTypeStrategy(explicit(DataTypes.BIGINT()))
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2279]: There are no new strategies here. `explicit` strategy is already covered by tests. Also I am sure we have IT tests for those functions that so far worked with the old stack and now they work with new stack.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        final boolean isPrecedingRowWindow =
+                isRowWindow(callContext, precedingTypeRoot, FollowingPreceding.PRECEDING);
+        final boolean isFollowingRowWindow =
+                isRowWindow(callContext, followingTypeRoot, FollowingPreceding.FOLLOWING);
+
+        if (isPrecedingRowWindow && !isFollowingRowWindow
+                || !isPrecedingRowWindow && isFollowingRowWindow) {
+            return callContext.fail(
+                    throwOnFailure, "Preceding and following must be of same interval type.");
```
Comment (jnh5y) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java:97]: If this is what we are checking, would it be more direct to check the types directly?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/OverConvertRule.java [hunk 5]
```diff
@@ -182,57 +191,59 @@ private RexNode createCollation(
...
+        if (bound instanceof ValueLiteralExpression) {
+            final ValueLiteralExpression literal = (ValueLiteralExpression) bound;
+            return literal.getValueAs(OverWindowRange.class)
+                    .map(r -> createSymbolBound(r, sqlKind))
+                    .orElseGet(
+                            () ->
+                                    createLiteralBound(
+                                            context, (ValueLiteralExpression) bound, sqlKind));
+        } else {
+            throw new TableException("Unexpected expression: " + bound);
```
Comment (jnh5y) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/OverConvertRule.java:203]: Have we tested this case?

(I'm asking to get an idea how this is called.)

**Review by jnh5y — APPROVED** [review]
Overall, looks good to me!

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2084]: Got it.  I see it now; looks good.

[the remark this answers is not in the document]
Comment (liuyongvs) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2084]: yeap, Those were actually not really functions, it should be as symbols like order(asc/desc)

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                if (callContext.getArgumentValue(followingPreceding.pos, Long.class).get() <= 0) {
+                    return Optional.of(
+                            String.format(
+                                    "%s row interval must be larger than 0.",
+                                    followingPreceding.name));
+                }
+                break;
+            case INTERVAL_DAY_TIME:
+                if (callContext
+                                .getArgumentValue(followingPreceding.pos, BigDecimal.class)
```
Comment (liuyongvs) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java:150]: when is can be INTERVAL_DAY_TIME?

**Comment by liuyongvs** [comment]
LGTM, only left one minor comments

**Review by liuyongvs — APPROVED** [review]

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java (hunk 4 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java:97]: symbols have symbol types. You can say `2 rows preceding CURRENT_ROW`, but you cannot say `2 rows preceding CURRENT_RANGE`.

Range = time, ROW = number of rows.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/OverConvertRule.java (hunk 5 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/OverConvertRule.java:203]: No, we have not. It should not be possible to reach this line. This is a) precaution b) solves checkstyle warning

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java (hunk 6 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java:150]: When you say e.g. `2.minutes preceding CURRENT_RANGE`
